TraderJoe-Bots/Large-Swap/src/agent.ts

```diff
@@ -0,0 +1,98 @@
+import { Finding, HandleTransaction, TransactionEvent, getEthersProvider } from "forta-agent";
+import { providers, BigNumber } from "ethers";
+import NetworkManager from "./network";
+import NetworkData from "./network";
+import { SWAP_ABI, PAIR_ABI, create2Pair } from "./utils";
+import { createFinding } from "./findings";
+import { MulticallContract, MulticallProvider } from "./multicall2";
+
+const THRESHOLD_PERCENTAGE: number = 20;
+
+const networkManager: NetworkData = new NetworkManager();
+
+export const provideInitialize = (provider: providers.Provider) => async () => {
+  const { chainId } = await provider.getNetwork();
+  networkManager.setNetwork(chainId);
+};
+
+export function provideHandleTransaction(
+  provider: providers.Provider,
+  networkManager: NetworkData,
+  thresholdPercentage: number
+): HandleTransaction {
+  return async (txEvent: TransactionEvent): Promise<Finding[]> => {
+    const findings: Finding[] = [];
+
+    await Promise.all(
+      // Filter for any `Swap` event emissions
+      txEvent.filterLog(SWAP_ABI).map(async (log) => {
+        // Assign the `Swap` arguments to variables
+        const [, amount0In, amount1In, amount0Out, amount1Out] = log.args;
+
+        // Create multicall2 instance based on
+        // the current provider's chain id
+        const multicall2: MulticallProvider = new MulticallProvider(
+          provider,
+          networkManager.multicall2Data,
+          networkManager.chainId
+        );
+
+        // Create pair contract at emitting address
+        // and fetch `token0` and `token1`
+        const pairContract = new MulticallContract(log.address, PAIR_ABI);
+        const tokenResults = await multicall2.all([pairContract.token0(), pairContract.token1()], txEvent.blockNumber);
```

Comment:
Maybe this multicall here could be on `txEvent.blockNumber - 1`, then it could also include the `getReserves` call.